sdk/core/core-http/src/policies/bearerTokenAuthenticationPolicy.ts

```diff
@@ -83,18 +85,37 @@ export class BearerTokenAuthenticationPolicy extends BaseRequestPolicy {
         spanOptions: webResource.spanOptions
       }
     });
-    webResource.headers.set(Constants.HeaderConstants.AUTHORIZATION, `Bearer ${token}`);
+    if (token) {
+      webResource.headers.set(Constants.HeaderConstants.AUTHORIZATION, `Bearer ${token}`);
+    }
     return this._nextPolicy.sendRequest(webResource);
   }
 
   private async getToken(options: GetTokenOptions): Promise<string | undefined> {
-    // We reset the cached token some before it expires,
+    // We reset the cached token in a time window before it expires,
     // after that point, we retry the refresh of the token only if the token refresher is ready.
+
     let token = this.tokenCache.getCachedToken();
-    if (!token && this.tokenRefresher.isReady()) {
-      token = await this.tokenRefresher.refresh(options);
-      this.tokenCache.setCachedToken(token);
+    let readyToRefresh = !token && this.tokenRefresher.isReady();
+    let refreshPromise: Promise<AccessToken | undefined> | undefined;
+    if (readyToRefresh) {
+      refreshPromise = this.tokenRefresher.refresh(options);
+    }
+
+    // If we have a local copy of the token, we use it.
+    if (this.token) {
+      token = this.token;
     }
+
+    // If we don't have a token, or if the token has expired, we wait for the refresh promise
+    if (!token || token.expiresOnTimestamp < Date.now()) {
```

Comment:
Should we add a buffer of time here? If the token is within 100ms of expiring, then there's a decent chance we won't get the request out the door and through the entire pipeline to validation at the service before it expires. Maybe we can say "if the token is within a second of expiring, then we wait."

Reply:
As long as the refresh window is not < 100ms, then there's no need for a buffer. We currently we default the refresh window to 2 minutes, and the refresh interval to 30 seconds.